airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/source.py

```diff
@@ -0,0 +1,516 @@
+        response_json = None
+        try:
+            response = requests.request(method="POST", url=self.token_refresh_endpoint, params=self.get_refresh_request_params())
+
+            response_json = response.json()
+            response.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            if response_json and "error" in response_json:
+                raise Exception(
+                    "Error refreshing access token {}. Error: {}; Error details: {}; Exception: {}".format(
+                        response_json, response_json["error"], response_json["error_description"], e
+                    )
+                ) from e
+            raise Exception(f"Error refreshing access token: {e}") from e
+        else:
+            return response_json["access_token"], response_json["expires_in"]
+
+    def get_refresh_request_params(self) -> Mapping[str, any]:
+        """
+        Sign the JWT with RSA-256 using the private key found in service account JSON file.
+        """
+        token_lifetime = 3600  # token lifetime is 1 hour
+
+        issued_at = time.time()
+        expiration_time = issued_at + token_lifetime
+
+        payload = {
+            "iss": self.client_email,
+            "sub": self.client_email,
+            "scope": self.scope,
+            "aud": self.token_refresh_endpoint,
+            "iat": issued_at,
+            "exp": expiration_time,
+        }
+
+        headers = {"kid": self.client_id}
+
+        signed_jwt = jwt.encode(payload, self.client_secret, headers=headers, algorithm="RS256")
+
+        return {"grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer", "assertion": str(signed_jwt)}
+
+
+class SourceGoogleAnalyticsV4(AbstractSource):
+    """Google Analytics lets you analyze data about customer engagement with your website or application."""
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        try:
+            url = f"{GoogleAnalyticsV4TypesList.url_base}"
+
+            authenticator = GoogleAnalyticsOauth2Authenticator(config)
+
+            session = requests.get(url, headers=authenticator.get_auth_header())
+            session.raise_for_status()
+
+            custom_reports = config.get("custom_reports")
+            if custom_reports:
+                json.loads(custom_reports)
+            return True, None
+        except (requests.exceptions.RequestException, ValueError) as e:
+            if e == ValueError:
+                logger.error(f"Invalid custom reports json structure.")
+            return False, e
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        streams: List[GoogleAnalyticsV4Stream] = []
+
+        authenticator = GoogleAnalyticsOauth2Authenticator(config)
+
+        config["authenticator"] = authenticator
+        config["ga_streams"] = json.loads(pkgutil.get_data("source_google_analytics_v4", "defaults/default_reports.json")) + json.loads(
+            config["custom_reports"]
```

Comment:
Let's say user provides valid JSON but content is invalid for the API. 

How will that failure look for the user? It should be clear that the failure was due to their invalid custom report JSON.

Reply:
Validation `custom_reports`  values handled `lookup_data_type` method.